File: quad_multi/cython_c/quad_openmp.py
from quad_openmp_wrapper import Integrator

I = Integrator()

jobs = [
        (0, 0.0, 10.0, 0.49936338107645674464, 1000000000),
        (1, -4.0, 4.0, 0.49936338107645674464, 1000000000)
       ]

for job in jobs:
    func, a, b, exact, n = job
    I.set_selection(func)
    I.set_parameters(a=a, b=b, exact=exact, n=n)

    print()
    print("QUAD_OPENMP:")
    print("  Python -> Cython -> C version")
    print("  Use OpenMP for parallel execution.")
    print("  Estimate the integral of f(x) from A to B.")
    print("  {}.".format(I.get_selection_repr()))
    print()
    print("  A        = {}".format(a))
    print("  B        = {}".format(b))
    print("  N        = {}".format(n))
    print("  Exact    = {}".format(exact))
    #print("  Exact    = {}%24.16f", exact)

    total, error, wtime = I.integrate()

    print()
    print("  Estimate = {}".format(total))
    print("  Error    = {}".format(error))
    print("  Time     = {}".format(wtime))

    print()
    print("QUAD_OPENMP:")
    print("  Normal end of execution.")
    print()

File: quad_multi/cython_c/quad_openmp.c
#include <math.h>
#include <omp.h>

int quad(double (*f)(double), double* total, double* error,
         double* wtime, double a, double b, double exact, int n) {
  int i;
  double t;
  double x;

  *wtime = omp_get_wtime ( );

  t = 0.0;

# pragma omp parallel shared ( a, b, n ) private ( i, x ) 

# pragma omp for reduction ( + : t )

  for ( i = 0; i < n; i++ ) {
    x = ( ( double ) ( n - i - 1 ) * a + ( double ) ( i ) * b ) / ( double ) ( n - 1 );
    t = t + f ( x );
  }

  *wtime = omp_get_wtime ( ) - *wtime;
  *total = ( b - a ) * t / ( double ) n;
  *error = fabs ( *total - exact );

  return 0;
}

